Cleaning up registered services

I am interested in dynamically loading and unloading components.
This works nice for standard components with `d:loadComponent()` and `d:unloadComponent()`, but when a `ROSServiceService` was loaded into a component, the service proxy seems to not "unadvertise" correctly on component unload and throws errors if the same service "topic" is used.

For instance, I added some controller manager orocos_services to a component by following this example https://github.com/jbohren/conman/blob/master/conman_ros/src/ros_interface_service.cpp  
This orocos_service uses the `ROSServiceService` to connect, meaning that it advertises a ros service through a `NodeHandle`.

However on unload, there is no disconnect in `ROSServiceService` and the `NodeHandle` apparently does not shutdown correctly so that I get errors if I redeploy the same component+orocos_service.

```
[ERROR] [1464168075.512551730, 1005.620000000]: Tried to advertise a service that is already advertised in this node [/ra/controller_manager/list_controller_types]
```

even if the component does not appear anymore (neither in `rttlib.stat()`  nor in `d:getPeers()`)

You might argue that a controller manager should always be there even in the case of dynamic deployment (conman does that by managing the list of peers and such), but the problem goes beyond the functionality of a controller manager, any orocos_service could create a `ROSServiceService` that at some point might needs to be cleaned-up for redeployment later on when using dynamic deployment. 

What would be a correct way to unadvertise the service ? at which level should it happen ? I could then try to modify the `rtt_roscomm` to handle that and create a PR.

--- a/rtt_roscomm/include/rtt_roscomm/rtt_rosservice_proxy.h
+++ b/rtt_roscomm/include/rtt_roscomm/rtt_rosservice_proxy.h
@@ -11,6 +11,7 @@ class ROSServiceProxyBase
 {
 public:
   ROSServiceProxyBase(const std::string &service_name) : service_name_(service_name) { }
+  virtual ~ROSServiceProxyBase() { }
   //! Get the name of the ROS service
   const std::string& getServiceName() const { return service_name_; }
 private:

--- a/rtt_roscomm/include/rtt_roscomm/rtt_rosservice_proxy.h
+++ b/rtt_roscomm/include/rtt_roscomm/rtt_rosservice_proxy.h
@@ -67,6 +67,12 @@ public:
         this);
   }
 
+  ~ROSServiceServerProxy()
+  {
+    // Clean-up advertized ROS services
+    server_.shutdown();     
+  }
+
 private:
   
   //! The callback called by the ROS service server when this service is invoked

--- a/rtt_roscomm/src/rtt_rosservice_service.cpp
+++ b/rtt_roscomm/src/rtt_rosservice_service.cpp
@@ -38,6 +38,22 @@ public:
     get_service_factory = rosservice_registry_->getOperation("getServiceFactory");
   }
 
+  ~ROSServiceService()
+  {
+    // Cleanup registered ROS services and clients
+    std::map<std::string, ROSServiceServerProxyBase*>::iterator iter_s;
+    for(iter_s=server_proxies_.begin(); iter_s != server_proxies_.end(); iter_s++)
+    {
+      delete iter_s->second;
+    }
+
+    std::map<std::string, ROSServiceClientProxyBase*>::iterator iter_c;
+    for(iter_c=client_proxies_.begin(); iter_c != client_proxies_.end(); iter_c++)
+    {
+      delete iter_c->second;
+    }
+  }
+
   //! Get an RTT operation caller from a string identifier
   RTT::base::OperationCallerBaseInvoker* get_owner_operation_caller(const std::string rtt_uri)
   {

